SwiftCompilerSources/Sources/Basic/SourceLoc.swift

```diff
@@ -0,0 +1,73 @@
+//===--- SourceLoc.swift - SourceLoc bridiging utilities ------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public struct SourceLoc {
+  public var pointer: UnsafePointer<UInt8>
+
+  public init?(pointer: UnsafePointer<UInt8>?) {
+    guard let pointer = pointer else {
+      return nil
+    }
+    self.pointer = pointer
+  }
+
+  public init?(pointer: UnsafePointer<CChar>?) {
+    guard let pointer = pointer else {
+      return nil
+    }
+    self.pointer = UnsafeRawPointer(pointer).assumingMemoryBound(to: UInt8.self)
+  }
+
+  public init?(bridged: BridgedSourceLoc) {
+    guard let pointer = bridged.pointer else {
+      return nil
+    }
+    self.init(pointer: pointer)
+  }
+
+  public var bridged: BridgedSourceLoc {
+    .init(pointer: pointer)
+  }
+}
+
+extension Optional where Wrapped == SourceLoc {
+  public var bridged: BridgedSourceLoc {
+    self?.bridged ?? .init(pointer: nil)
+  }
+}
+
+public struct CharSourceRange {
+  public var start: SourceLoc
+  public var byteLength: Int
```

Comment:
Why are these not `let`?

Reply:
In general, if a `struct` provides a member-wise initializer, there is no reason to make any of its property `let`. Even if we make these `let`, others can

```swift
extension CharSourceRange {
  mutating func setByteLength(val: Int) {
    self = .init(start: start, byteLength: value)
  }
}
```

Reply:
Given this is Swift vs. C++, it seems to me like an antipattern to expose opportunities for mutating public storage directly on an instance when it is not intended as publicly mutable, in terms of how error-prone it may happen to be in imperative code and when maintaining state integrity. In the extension it would at least be an "atomic" mutation. This is a personal opinion though.

Reply:
I agree with @AnthonyLatsis here. Making properties `let` enforces a more stateless design of the code, which is good to keep the code simple. If a property is `var`, it adds unnecessary complexity.

Reply:
OK, I changed them to `let`